### PR TITLE
Download job data in advance, upload results in background

### LINKED ISSUE #43: Download inputs for queued jobs in the background

### LINKED ISSUE #44: Upload outputs of finished jobs in the background

### LINKED ISSUE #59: Make shepherd robust to io fails
A number of operations in `_listen` co-routine may fail, e.g.: 
https://github.com/iterait/shepherd/blob/ede07cfa4ca68fa6b325aeca1e1f7837b5c9b9bb/shepherd/shepherd/shepherd.py#L297 
which effectively kills the co-routine and the shepherd is unable to recover from it.

Shepherd should be able to deal with these cases (and report them in the logs ofc).

This can be reproduced by running two `shepherd`s with the same working directory.

### PR DESCRIPTION
**Pull request by janbuchar**
closes #43
closes #44
closes #59

### CODE REVIEW

File: shepherd/shepherd/shepherd.py [hunk 1]
```diff
@@ -188,65 +206,105 @@ def _slaughter_sheep(self, sheep_id: str) -> None:
...
+        :param job_id: The job for which we are preparing the directory
+        """
+        logging.info('Preparing working directory for job `%s` on `%s`', job_id, sheep.id)
+        working_directory = create_clean_dir(path.join(sheep.sheep_data_root, job_id))
+        await self._storage.pull_job_data(job_id, working_directory)
+        create_clean_dir(path.join(working_directory, OUTPUT_DIR))
+
+    async def _reconfigure_sheep(self, sheep: BaseSheep, job_id: str, model: ModelModel) -> bool:
+        """
+        Slaughter a sheep and resurrect it with a different model.
```
Comment (blazekadam) [inline shepherd/shepherd/shepherd.py]: Whoa a resurrection 👍

**Review by blazekadam — COMMENTED** [review]
This looks good so far and the tests are passing but I will need to think a bit more about it. I am afraid of additional edge cases and race conditions...

**Comment by Teyras** [comment]
@blazekadam Sure, take your time - this is nice to have, but not essential by any means.

File: shepherd/shepherd/shepherd.py [hunk 2]
```diff
@@ -88,6 +94,8 @@ def __init__(self,
         self._listener = asyncio.create_task(self._listen())
         self._health_checker = asyncio.create_task(self._shepherd_health_check())
         self._job_status_update_queue = TaskQueue(worker_count=1)
+        self._prepare_job_dir_queue = TaskQueue(worker_count=8)
```
Comment (blazekadam) [inline shepherd/shepherd/shepherd.py]: What an arbitrary constant?

File: shepherd/shepherd/shepherd.py [hunk 3]
```diff
@@ -166,15 +178,21 @@ def _slaughter_sheep(self, sheep_id: str) -> None:
...
         while True:
             await asyncio.sleep(1)
-            sheep = self._get_sheep(sheep_id)
             try:
                 if not sheep.running:
                     for job_id in sheep.in_progress:
                         # clean-up the working directory
                         shutil.rmtree(path.join(self._get_sheep(sheep_id).sheep_data_root, job_id))
 
+                        # if necessary, cancel input downloads
```
Comment (blazekadam) [inline shepherd/shepherd/shepherd.py]: The jobs should be imo canceled before the `rmtree`

File: shepherd/shepherd/shepherd.py [hunk 4]
```diff
@@ -188,65 +206,105 @@ def _slaughter_sheep(self, sheep_id: str) -> None:
...
+        """
+        Slaughter a sheep and resurrect it with a different model.
+
+        :param sheep: the sheep to be reconfigured
+        :param job_id: ID of the job that prompted the reconfiguration
+        :param model: the new model
+        :return: whether or not the reconfiguration succeeded
+        """
+
+        # we need to wait for the in-progress jobs which might already be in the socket
```
Comment (blazekadam) [inline shepherd/shepherd/shepherd.py]: What a respectful way to slaughter a sheep :o

File: shepherd/shepherd/shepherd.py [hunk 5]
```diff
@@ -188,65 +206,105 @@ def _slaughter_sheep(self, sheep_id: str) -> None:
...
+        async with self.job_done_condition:
+            await self.job_done_condition.wait_for(lambda: len(sheep.in_progress) == 0)
+
+        self._slaughter_sheep(sheep.id)
+
+        try:
+            self._start_sheep(sheep.id, model.name, model.version)
+            return True
+        except SheepConfigurationError as sce:
```
Comment (blazekadam) [inline shepherd/shepherd/shepherd.py]: The `except` clauses looks quite similar, can we join them or is there a good reason to keep them separated?

File: shepherd/shepherd/shepherd.py [hunk 6]
```diff
@@ -188,65 +206,105 @@ def _slaughter_sheep(self, sheep_id: str) -> None:
...
-                        'exception_type': str(type(ex)),
-                        'exception_traceback': str(traceback.format_tb(ex.__traceback__))
-                    })
-
-                    await self._report_job_failed(job_id, error, sheep)
-                    logging.exception("Error encountered when starting sheep `%s` for job `%s`", sheep_id, job_id)
+                    await self._prepare_job_dir_futures[job_id]
+                    del self._prepare_job_dir_futures[job_id]
+                except KeyError:
+                    logging.critical(f"Task that should have pulled data for job `{job_id}` mysteriously disappeared")
```
Comment (blazekadam) [inline shepherd/shepherd/shepherd.py]: This may happen anytime a sheep is purged with `health_check`, I do not think that it deserves the `critical` level...

File: shepherd/shepherd/shepherd.py [hunk 7]
```diff
@@ -188,65 +206,105 @@ def _slaughter_sheep(self, sheep_id: str) -> None:
...
+                        continue
+
+                # send the InputMessage to the sheep
+                sheep.in_progress.add(job_id)
+                logging.info('Sending InputMessage for job `%s` on `%s`', job_id, sheep_id)
+                await Messenger.send(
+                    sheep.socket, InputMessage(dict(job_id=job_id, io_data_root=sheep.sheep_data_root))
+                )
+
+                # notify the queue that the task is done
```
Comment (blazekadam) [inline shepherd/shepherd/shepherd.py]: This is weird wording (line 303 as well), the job is certainly not done, it was just sent for processing

File: shepherd/shepherd/shepherd.py [hunk 8]
```diff
@@ -266,6 +324,44 @@ def _slaughter_sheep(self, sheep_id: str) -> None:
...
+        job_id = message.job_id
+
+        # clean-up the working directory and upload the results
+        working_directory = path.join(self._get_sheep(sheep.id).sheep_data_root, job_id)
+        await self._storage.push_job_data(job_id, working_directory)
+        shutil.rmtree(working_directory)
+
+        # save the done/error file
+        if isinstance(message, DoneMessage):
+            status = self._job_status.pop(job_id)
```
Comment (blazekadam) [inline shepherd/shepherd/shepherd.py]: Same as the line 356, can we put it before the `if-else`?

File: shepherd/shepherd/shepherd.py [hunk 9]
```diff
@@ -266,6 +324,44 @@ def _slaughter_sheep(self, sheep_id: str) -> None:
...
+            status.status = JobStatus.DONE
```
Comment (blazekadam) [inline shepherd/shepherd/shepherd.py]: I do not like how `DoneMessage` is processed in place while `ErrorMessage` is mostly handled with `self._report_job_failed`. I.e., we do quite the same at two different places and two different depths...

**Review by blazekadam — CHANGES_REQUESTED** [review]

**Comment by blazekadam** [comment]
Nothing really waits for this, right?

**Comment by Teyras** [comment]
That's right
